### PR TITLE
Suppress some errors when speculatively resolving initializer calls

### LINKED ISSUE #22451: Unexpected error with array type alias with non-default initializeable type
### Summary of Problem

I would expect the following code to work because no array element is default initialized. But, instead it gives a compilation error. A workaround is to create an initializer to support default initialization (in this case, with a `type T` formal).

Looking within the compiler, it looks like the error is being emitted while resolving a call to `isDefaultInitializableType`.

### Steps to Reproduce

**Source Code:**

```chapel
record R {
  type T;
  var x: T;
  proc init(type T, arg: T) {
    this.T = T;
    this.x = arg;
  }
  proc init(arg: ?t) {
    this.T = t;
    this.x = arg;
  }
}

type D = [0..#2] R(real);

proc main() {
  var x: D = [new R(1.0), new R(2.0)];
  writeln(x);
}
```

**Compile command:**

```
$ chpl bb.chpl  --no-devel
bb.chpl:14: error: unresolved call 'R.init(T=type real(64))'
bb.chpl:4: note: this candidate did not match: R.init(type T, arg: T)
bb.chpl:14: note: because call does not supply enough arguments
bb.chpl:4: note: it is missing a value for formal 'arg'
bb.chpl:14: note: other candidates are:
bb.chpl:8: note:   R.init(arg: ?t)
```

**Context:**

I'm filing this issue on behalf of @lucaferranti.

This came about in the context of changing a type from being concrete to generic (along with #22450).

**Associated Future Test(s):**
TODO

### Configuration Information

Fails on `main`

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
Resolves #22451.

Adjust resolution so that it is possible to suppress some errors when resolving initializers. This enables speculative resolution of default initializers for features like `isDefaultInitializable`. Now resolution will not emit a fatal error if no candidates exist for a default initializer.

The old compiler has a framework for trying to resolve calls without emitting errors, but initializer resolution was not integrated into this framework. Add a `bool` argument `emitCallResolutionErrors` to `resolveInitializer` and `resolveInitCall` which suppresses some errors if it is true.

This is not taken as a guarantee to suppress all errors (in particular, if an initializer body is resolved, even during the course of candidate selection, errors will still be emitted). However, this is fine, because we will gain well-implemented speculative resolution when the `dyno` resolver becomes the default.

TESTING

- [x] `linux64`, `standard`

Reviewed by @lydia-duncan. Thanks!

### CODE REVIEW

File: compiler/resolution/initializerResolution.cpp [hunk 1]
```diff
@@ -554,33 +559,39 @@ static void resolveInitCall(CallExpr* call, AggregateType* newExprAlias, bool fo
...
+            if (candidates.n == 0) {
+              printResolutionErrorUnresolved(info, mostApplicable);
+
+              USR_STOP();
+            } else {
+              printResolutionErrorAmbiguous (info, candidates);
+            }
           } else {
-            printResolutionErrorAmbiguous (info, candidates);
+            INT_ASSERT(tryingToResolve());
```
Comment (lydia-duncan) [inline compiler/resolution/initializerResolution.cpp:585]: I'm not sure how I feel about this - if we shouldn't emit errors, why is turning this into an internal compiler error okay?

**Review by lydia-duncan — COMMENTED** [review]
I think my main concern with this adjustment is that initializer calls are different from normal function calls - there are situations where the arguments determine what the eventual type will be and situations where it's determined by the initializer itself, and the `this` argument has to be treated specially as a result.  So if we've gotten ourselves into a situation where the call can't be resolved, it generally needs to be handled through the initializer-specific machinery to give a reasonable error message for the user.

That said I (perhaps egotistically :P) have more confidence in the presence of initializer tests that would detect this than I do for other parts of our language, so I'm somewhat reassured that the only testing change necessary was adding a new test.

I have one comment in the code for us to discuss, but that's about it otherwise

File: compiler/resolution/initializerResolution.cpp (hunk 1 above)
Comment (dlongnecke-cray) [reply compiler/resolution/initializerResolution.cpp:585]: Good point, this assertion is a holdover from before the `bool` formal.

**Comment by dlongnecke-cray** [comment]
Yeah, in general if an initializer candidate is _found_, all the usual errors will be emitted. Perhaps I should rename the `bool` to something like `emitErrorsIfNoCandidatesFound`?

**Comment by dlongnecke-cray** [comment]
I've removed the internal assertion and renamed the `bool` to `emitCallResolutionErrors`, as well as added a comment.

**Pull request merged**